src/mongocrypt-crypto-private.h

```diff
@@ -44,116 +44,50 @@ typedef struct {
-                                    const _mongocrypt_buffer_t *associated_data,
-                                    const _mongocrypt_buffer_t *key,
-                                    const _mongocrypt_buffer_t *ciphertext,
-                                    _mongocrypt_buffer_t *plaintext,
-                                    uint32_t *bytes_written,
-                                    mongocrypt_status_t *status)
-   MONGOCRYPT_WARN_UNUSED_RESULT;
-
-/* _mongocrypt_fle2_do_encryption does non-AEAD encryption.
- * @key is expected to be only an encryption key of size MONGOCRYPT_ENC_KEY_LEN.
- * Note: Attempting to encrypt a 0 length plaintext is an error.
- */
-bool
-_mongocrypt_fle2_do_encryption (_mongocrypt_crypto_t *crypto,
-                                const _mongocrypt_buffer_t *iv,
-                                const _mongocrypt_buffer_t *key,
-                                const _mongocrypt_buffer_t *plaintext,
-                                _mongocrypt_buffer_t *ciphertext,
-                                uint32_t *bytes_written,
-                                mongocrypt_status_t *status)
-   MONGOCRYPT_WARN_UNUSED_RESULT;
+typedef uint32_t (*_mongocrypt_ciphertextlen_fn) (uint32_t plaintext_len,
+                                                  mongocrypt_status_t *status);
+typedef uint32_t (*_mongocrypt_plaintextlen_fn) (uint32_t ciphertext_len,
+                                                 mongocrypt_status_t *status);
+typedef bool (*_mongocrypt_do_encryption_fn) (
+   _mongocrypt_crypto_t *crypto,
+   const _mongocrypt_buffer_t *iv,
+   const _mongocrypt_buffer_t *associated_data,
+   const _mongocrypt_buffer_t *key,
+   const _mongocrypt_buffer_t *plaintext,
+   _mongocrypt_buffer_t *ciphertext,
+   uint32_t *bytes_written,
+   mongocrypt_status_t *status) MONGOCRYPT_WARN_UNUSED_RESULT;
+typedef bool (*_mongocrypt_do_decryption_fn) (
+   _mongocrypt_crypto_t *crypto,
+   const _mongocrypt_buffer_t *associated_data,
+   const _mongocrypt_buffer_t *key,
+   const _mongocrypt_buffer_t *ciphertext,
+   _mongocrypt_buffer_t *plaintext,
+   uint32_t *bytes_written,
+   mongocrypt_status_t *status) MONGOCRYPT_WARN_UNUSED_RESULT;
 
-/* _mongocrypt_fle2_do_decryption does non-AEAD decryption.
- * @key is expected to be only an encryption key of size MONGOCRYPT_ENC_KEY_LEN.
+/**
+ * Defines the application layer protocol to use when
+ * encrypting client data values.
  */
-bool
-_mongocrypt_fle2_do_decryption (_mongocrypt_crypto_t *crypto,
-                                const _mongocrypt_buffer_t *key,
-                                const _mongocrypt_buffer_t *ciphertext,
-                                _mongocrypt_buffer_t *plaintext,
-                                uint32_t *bytes_written,
-                                mongocrypt_status_t *status)
-   MONGOCRYPT_WARN_UNUSED_RESULT;
+typedef struct {
+   _mongocrypt_ciphertextlen_fn ciphertext_len;
```

Comment:
nit - I might prefix `ciphertext_len` and `plaintext_len` with a verb like `get` to make it more obvious that they are functions instead just variables.

When I was reading code, like

`fle2->plainext_len(...) `

my mind kept asking, `why is a variable being called like a function`